### ISSUE TITLE
Support for PRELIM auth connection mode?

### ISSUE THREAD

**Issue by kurt-google** [go-goracle/goracle#127]
ODPI supports `DPI_MODE_AUTH_PRELIM` when connecting to issue privileged or maintenance operations (startup/shutdown). ref: https://oracle.github.io/odpi/doc/enums/dpiAuthMode.html#dpiauthmode

Once implemented the oracle specific driver extensions could have `dpiConn_startupDatabase` support and `dpiConn_shutdownDatabase` (or better a wrapped shutdown implementation so users dont need to call shutdown twice per the documentation ref: https://docs.oracle.com/en/database/oracle/oracle-database/18/lnoci/database-startup-and-shutdown.html#GUID-44B24F65-8C24-4DF3-8FBF-B896A4D6F3F3

Has this feature been considered previously? Is it possible to add this in as an additional parameter to the jdbc style url connection string? Currently no go driver appears to implement these features so it would be great to have this so a single go driver could run all maintenance tasks. (or even build a usable alternative to sqlplus)

**Comment by kurt-google** [comment]
I have not tested yet as no drivers support it, but with sqlplus you can issue CREATE PFILE while the database is offline through this mode.

Shutdown does not need to be issued from this mode per the docs. However if your database is hung you can issue a SHUTDOWN ABORT from this mode or query SGA without ever establishing a connection. (sqlplus for example offers the undocumented -prelim flag for this functionality)

**Comment by kurt-google** [comment]
Startup also requires two connections, one in PRELIM to start the oracle processes (`dpiConn_startupDatabase`) after which you need to establish a default connection to issue the typical ALTER DATABASE MOUNT/OPEN (also in docs). So a single prelim connection is only good for starting processes which doesn't present as nicely as oracle's `STARTUP MODE`.

Wrapping these would be a neat feature to add for the general startup/shutdown cases but also not really required at the driver level. And subverts the purpose of exposing prelim auth to advanced users.

**Comment by kurt-google** [comment]
> So, what can be done with PRELIM, and what can't?
An excellent question, however this is a lightly documented area of oracle and I'm no expert.

Tested as using the ODPI/OCI are startup/shutdown/CREATE (S)PFILE. I tested ORADEBUG (another somewhat commonly used tool in prelim mode) but ORADEBUG appears to be implemented outside OCI.

Outside of these 3 (not including ORADEBUG) im not sure if any other functionality is available in this mode. But personally I use all 3 and would want any implementation to support them all.

**Comment by tgulacsi** [comment]
So, how should the API look like?

a) do nothing, just allow the prelim=1 mode in the connection string
b) if sysdba=1 or sysoper=1, then for such a connection, allow a `goracle.NewPrelim` call, which returns a new object (with inderlying connection wth prelim), with a Shutdown, Startup and a CreatePfile(serverSide bool) method.

a) is the easiest, but if you cannot use a PRELIM connection for anything else, then maybe b) is nicer.

What do you think about it?

**Comment by kurt-google** [comment]
a) was only a means to the end of exposing the startup/shutdown functions (they are not exposed as SQL in the prelim connection mode).

b) is what I would like the most, though i think re-checking oracle's own checks seems redundant. That session/object with an interface like `Startup(dpiStartupMode)`,`ShutdownAbort()`,`Exec(sql string)`

Reasonings:
Startup Restricted mode is useful for maintenance, startup Force mode is less useful especially with `ShutdownAbort()` available.
Shutdown modes other than ABORT should be issued from standard sessions
pfile/spfile sql docs: https://docs.oracle.com/cd/B28359_01/server.111/b28286/statements_6008.htm#SQLRF01308
https://docs.oracle.com/cd/B28359_01/server.111/b28286/statements_6016.htm#SQLRF01315
And as far as i can tell (without actually opening a prelim auth connection outside sqlplus), `OciStatementExec` is being issued to complete CREATE SPFILE/CREATE PFILE commands. And errors are being returned via the typical `OciErrorGet` process. So just leaving this available and oracle should return ORA-01012 for everything that isnt allowed.

Otherwise the `CreatePfile` (and probably separate `CreateSpfile` for visibility/clarity) needs to handle MEMORY sources, Default source (e.g. PFILE from SPFILE, or SPFILE from PFILE), and specific sources (CREATE SPFILE from PFILE='/u01/backup/somebackupSpfile.ora'). Where as Execute is already implemented.

**Comment by kurt-google** [comment]
I havnt checked yet but shutdown may be possible via typical SQL flows, otherwise eventually also a `shutdown()` on the _normal_ session should be exposed to allow for the typical `dpiConn_shutdownDatabase(DPI_MODE_SHUTDOWN_{DEFAULT,TRANSACTIONAL,IMMEDIATE})` to begin the shutdown and the final `dpiConn_shutdownDatabase(DPI_MODE_SHUTDOWN_FINAL)`

Separating these as SQL must be issued between shutdown attempts, or integrating the SQL calls into the wrapped function.  ref: odpi https://oracle.github.io/odpi/doc/functions/dpiConn.html#c.dpiConn_shutdownDatabase
`After calling this function the first time, the SQL statements “alter database close normal” and “alter database dismount” must be executed. `

**Comment by tgulacsi** [comment]
Ok, it's still not ok: if you want to issue some commands on the PRELIM connection, with this API you cannot do it.
Maybe we should return the prelim'd connection with Conn.NewPrelim ?

**Comment by cjbj** [comment]
FWIW cx_Oracle has [startup](https://github.com/oracle/python-cx_Oracle/blob/7.0/src/cxoConnection.c#L1471)  and [shutdown](https://github.com/oracle/python-cx_Oracle/blob/7.0/src/cxoConnection.c#L1442).  Also see [PRELIM_AUTH doc](https://cx-oracle.readthedocs.io/en/latest/module.html?highlight=startup#cx_Oracle.PRELIM_AUTH).

**Comment by kurt-google** [comment]
I think as written it is good except for the shutdown forcing prelim connections. I will try and pay with it later.

As written I can pass prelim to the primary connection and issue exec that way, no need to get the Conn.NewConn right?

**Comment by tgulacsi** [comment]
Right. Reading cx_Oracle's documentation (thanks, @cjbj !), now I think the simplest and most flexible solution would be offering a Shutdown and Startup method for the connection, and allowing PRELIM_AUTH connection, separately.

This way all Shutdown can be issued on either plain or prelim connection, and independently you can use the prelim connection as you wish.

Less magic, simpler solution.

**Comment by tgulacsi** [comment]
@kurt-google could you test it? Shall I merge it?
Or any other suggestions, modifications?

**Comment by kurt-google** [comment]
My cursory review of the commits seemed good. I'll make some time to test it out, and hopefully that will give anyone watching time to comment.

**Comment by kurt-google** [comment]
Hey sorry for the slow turn around, but I've gotten to testing and am running into a couple preliminary issues.

Trying to connect as sysdba to the local database parses incorrectly it appears. A connection string of "/ as sysdba" ends up hitting
https://github.com/go-goracle/goracle/blob/f5367efc9422cda1720a92c119da7ff2d107e252/drv.go#L504
 and not using external auth. Which seems like a bug as it should register SYSDBA and create a connection instead of a pool.

Trying to use a connection string of `oracle://SID?sysdba=1` errors out in what seems like the appropriate place
https://github.com/go-goracle/goracle/blob/f5367efc9422cda1720a92c119da7ff2d107e252/drv.go#L469
The connection parameters appear correct with `ExternAuth=1`, however I get an `ORA-01017: invalid username/password; logon denied` which doesnt happen when using for example, `sqlplus`.

Attempting a prelim connection with `oracle://SID?sysdba=1&prelim=1` fails similarly with `ORA-01017`

Will try again with a password but extern auth is required for prelim connections I believe, since the database wont be online to authenticate you you must rely on OS authentication.

**Comment by kurt-google** [comment]
https://github.com/go-goracle/goracle/blob/prelim/drv.go#L385
Switch/case makes these parameters exclusive, but `prelim` must be used with `sysdba` so any attempt to properly set `?sysdba=1&prelim=1` results in auth mode of only `sysdba`. This code should simply OR everything you have set together instead of only ORing in the first param that matches the switch.

**Comment by kurt-google** [comment]
Shutdown as `sys` works great! Example workflow might be something like:

```go
        oraDB, err := goracle.DriverConn(db)                                                                 
        if err != nil {                                                                                      
                return err                                                                                   
        }                                                                                                    
        log.Infof("Beginning shutdown %v", shutdownMode)                                                     
        if err = oraDB.Shutdown(shutdownModeEnum); err != nil {                                              
                return err                                                                                   
        }                                                                                                    
        // If we abort the shutdown process is over immediately.                                             
        if shutdownModeEnum == goracle.ShutdownAbort {                                                       
                return nil                                                                                   
        }                                                                                                    
                                                                                                             
        log.Info("Closing database")                                                                         
        if _, err = db.Exec("alter database close normal"); err != nil {                                     
                return err                                                                                   
        }                                                                                                    
        log.Info("Unmounting database")                                                                      
        if _, err = db.Exec("alter database dismount"); err != nil {                                         
                return err                                                                                   
        }                                                                                                    
        log.Info("Finishing shutdown")                                                                       
        if err = oraDB.Shutdown(goracle.ShutdownFinal); err != nil {                                         
                return err                                                                                   
        }                                                                                                    
        return nil
```

Unfortunately the issues with prelim prevent me from testing the other features. I will try and take a look at why it might be giving me invalid login issues when I am appropriately privileged. Maybe its a simple as fixing the sysdba overwriting prelim auth flags.

**Comment by kurt-google** [comment]
https://github.com/oracle/odpi/blob/master/src/dpiConn.c#L295

Suggests that setting a connection class by default may result in undesirable behavior. In my tests it appeared to cause the connection to attempt to allocated structures in the oracle shared memory location (which doesnt exist while oracle is offline, resulting in failure to connect in prelim auth). However even without a connectionClass I am still getting `ORA-01017`

**Comment by kurt-google** [comment]
Ah so I found the issue, when calling `dpiConn_create` passing `NULL` for the `connectString` is required to initiate a local connection (ref https://oracle.github.io/odpi/doc/functions/dpiConn.html#c.dpiConn_create). Earlier in `parseConnectionString` the driver artificially populates the SID from `$ORACLE_SID`. This env var should be left for the ODPI layer to interpret during local connections and the user should be allowed to pass an empty (`NULL`) connection string to initiate a local connection (like the `/ as sysdba` connection string in sqlplus, the target is empty to signify a local connection versus connecting through a Listener).

As written all connections are forced through the listener since the connection string is never `NULL`. Since the database is offline the listener reasonably either complains or gives the opaque `ORA-01017`.

Setting up a test where I simply passed in `nil` to `dpiConn_create` for the connection string and forced externAuth to 0 (ref: https://docs.oracle.com/en/database/oracle/oracle-database/18/lnoci/database-startup-and-shutdown.html#GUID-44B24F65-8C24-4DF3-8FBF-B896A4D6F3F3 during startup on the prelim auth they use `OCI_CRED_RDBMS` and not `OCI_CRED_EXTERN`), I was able to connect with arbitrary user/pass and initiate a successful startup.

For most of my use cases I prefer local connections as the tool will be running locally and this removes dependencies on TNS_ADMIN and possibly variable listener configurations per system. This also makes it behave more similarly to sqlplus which should help users. Documentation of dependence on ORACLE_HOME and ORACLE_SID for local connections might be worthwhile for people who are not already familiar with OCI/sqlplus.

**Comment by kurt-google** [comment]
My preference is to simply remove the code which populates `P.SID` based on the environment variables, and let the user specify a connection string without a SID/ConnectString so that it properly gets resolved internally in OCI to a local connection.

**Comment by kurt-google** [comment]
These changes to get startup working also enable `CREATE PFILE/SPFILE` via the normal sql.Query/Exec workflow. :tada:

**Comment by tgulacsi** [comment]
Ok, I've removed the SID magic, please try 023011a.

**Comment by kurt-google** [comment]
All tests passing on 023011a except `shutdown abort` on `prelim` auth, However ive never once used that.

Using a connection string like `oracle://?sysdba=1&prelim=1` for prelim and `oracle://?sysdba=1` for normal local sysdba connections works great (equivalent of `sqlplus / as sysdba`)

Example startup workflow
```go
        oraDB, err := goracle.DriverConn(db)                                                                 
        if err != nil {                                                                                      
                return err                                                                                   
        }                               
        log.Info("Starting database")
        if err := oraDB.Startup(startupModeEnum); err != nil {
                return err                   
        }                
        // You cannot alter database on the prelim_auth connection.
        // So open a new connection and complete startup, as Startup starts pmon.          
        db2, err := sql.Open("goracle", "oracle://?sysdba=1")
        if err != nil {   
                return err
        }                      
        defer db2.Close()
                                        
        log.Info("Mounting database")
        if _, err := db2.Exec("alter database mount"); err != nil {
                return err                   
        }                
        log.Info("Opening database")         
        if _, err := db2.Exec("alter database open"); err != nil {
                return err
        }         
        return nil
```

**Comment by kurt-google** [comment]
Thanks so much for all your work getting things patched up for these features!

**Comment by tgulacsi** [comment]
It's your thorough testing with detailed instruction that made this possible!

I'll merge it into a new version soon.

**Comment by tgulacsi** [comment]
It's been released as v2.11.0.

Thanks for your support!
